### PR TITLE
Nvfuser python API import fix

### PR DESCRIPTION
**Pull request by jjsjann123**
1. Having nvfuser python API import working with both devel and upstream;
2. Add environment variable to allow custom nvfuser code base to be built with upstream pytorch core.

### CODE REVIEW

**Comment by jjsjann123** [comment]
This would help our internal container nightly build to merge the nvfuser & pytorch container.

**Comment by davidberard98** [comment]
this will need a lint fix - maybe there's a way to skip that check for these import locations?

I'll give @malfet a chance to comment on the other change

**Comment by jjsjann123** [comment]
Ah, the python import is done lazily in pytorch and that seems to be ok... Let me patch these.

File: torch/_prims/nvfuser_executor.py [hunk 1]
```diff
@@ -20,11 +20,12 @@
 
 if torch.cuda.is_available():
     import nvfuser
-    from packaging.version import Version
+    from packaging.version import Version  # type: ignore[import]
```
Comment (crcrpar) [inline torch/_prims/nvfuser_executor.py:23]: I'm not sure about this repo's way of importing `packaging` module. at glance, https://github.com/pytorch/pytorch/blob/d2b82feb41552e93efdd8f876b7a938d9657a1d6/test/onnx/test_verification.py#L14 is the only one which directly imports the module w/o guard. The others seem to import it from `pkg_resources` or https://github.com/pytorch/pytorch/pull/71902

Comment (jjsjann123) [reply torch/_prims/nvfuser_executor.py:23]: Yeah I think upstream relies on pkg_package in absence of `packaging`, tho I don't know what the difference is. I think I can use the lazy import from torch_version.py to handle this instead.
I'm not sure about the dependency between that with prim. We'll see if it breaks anything :crossed_fingers: 

Will ping you once I have it updated.

Comment (jjsjann123) [reply torch/_prims/nvfuser_executor.py]: switched to torch.torch_version.Version instead. Local testing seems to be working fine.

**Comment by jjsjann123** [comment]
@davidberard98 Do we still have concern on this PR?

File: torch/_prims/nvfuser_executor.py [hunk 2]
```diff
@@ -19,12 +19,25 @@
...
+            FusionDefinition,
+            Tensor,
+        )
+    else:
+        from nvfuser._C import (  # type: ignore[import]
+            DataType,
+            Fusion,
+            FusionDefinition,
+            Tensor,
+        )
```
Comment (malfet) [inline torch/_prims/nvfuser_executor.py:40]:
```suggestion
    try:
        from nvfuser import (  # type: ignore[attr-defined, import]
            DataType,
            FusionDefinition,
            Tensor,
        )
        def create_fusion_definitinon():
            return FusionDefinition()
    except ImportError:
        from nvfuser._C import (  # type: ignore[import]
            DataType,
            Fusion,
            FusionDefinition,
            Tensor,
        )
        def create_fusion_definition():
            f = Fusion()
            return FusionDefinition(f)
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -1170,7 +1170,11 @@ if(NOT USE_CUDA AND NOT USE_ROCM)
 endif()
 
 if(BUILD_NVFUSER)
-  add_subdirectory(third_party/nvfuser)
+  if(DEFINED ENV{NVFUSER_SOURCE_DIR})
+    add_subdirectory($ENV{NVFUSER_SOURCE_DIR} nvfuser)
+  else()
+    add_subdirectory(third_party/nvfuser nvfuser)
+  endif()
```
Comment (malfet) [inline CMakeLists.txt:1177]: Is this a temporary workaround?

**Review by malfet — CHANGES_REQUESTED** [review]
Please use more pythonic paradigms(rather then version checks) to check whether old or new nvfuser is around

File: CMakeLists.txt (hunk 3 above)
Comment (jjsjann123) [reply CMakeLists.txt:1177]: Yes!

Before we get our submodule setup properly, we would hope to stick with upstream pytorch code base `https://github.com/pytorch/pytorch/` while using our development nvfuser branch for our internal nightly container.

File: torch/_prims/nvfuser_executor.py [hunk 4]
```diff
@@ -148,8 +164,14 @@ def make_nvfuser_fusion(gm: GraphModule, *nv_args_templates):
...
-    fusion = Fusion()
-    with FusionDefinition(fusion) as fd:
+    if nvfuser_version >= Version("0.0.1"):
+        fusion = FusionDefinition()
+        fd = fusion
+    else:
+        fusion = Fusion()
+        fd = FusionDefinition(fusion)
+
+    with fd:
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply torch/_prims/nvfuser_executor.py:174]: This one might need some patching after the change. I'll take it as-is and handle that afterwards.

**Comment by jjsjann123** [comment]
No idea what's wrong the the linter failing. Looks like infra issue. I'll restart it later.

**Comment by jjsjann123** [comment]
Patched a linter error that was missed. CI looks good from previous run otherwise.

**Comment by jjsjann123** [comment]
CI looks green~~ :partying_face: 

pinging @malfet to approve changes per request

**Comment by jjsjann123** [comment]
Bumping for update. cc'ing @malfet

**Review by davidberard98 — APPROVED** [review]
looks like @malfet's requested changes have been resolved

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**